dbms/src/Encryption/RateLimiter.cpp

```diff
@@ -338,10 +362,24 @@ bool ReadLimiter::canGrant([[maybe_unused]] Int64 bytes)
 
 void ReadLimiter::refillAndAlloc()
 {
-    if (available_balance < refill_balance_per_period)
+    // `available_balance` of `ReadLimiter` may be overdrawn.
+    if (available_balance < 0)
     {
-        available_balance += refill_balance_per_period;
+        // Limiter may not be called for a long time.
+        // During this time, limiter can be refilled at most `max_refill_times` times and covers some overdraft.
+        auto elapsed_duration = std::chrono::system_clock::now() - last_refill_time;
+        UInt64 elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed_duration).count();
+        // At least refill one time.
+        Int64 max_refill_times = std::max(elapsed_ms, refill_period_ms) / refill_period_ms;
+        Int64 max_refill_bytes = max_refill_times * refill_balance_per_period;
+        alloc_bytes += std::min(-available_balance, max_refill_bytes);
+        available_balance = std::min(available_balance + max_refill_bytes, refill_balance_per_period);
     }
+    else
+    {
+        available_balance = refill_balance_per_period;
```

Comment:
Does here fix the issue "The calculation of alloc_bytes is not right when the burst read traffic exceeds the limit and results in the I/O limiter-tuner not working well." ?